diff --git a/.gitignore b/.gitignore
index b0ae383..21bc34c 100644
--- a/.gitignore
+++ b/.gitignore
@@ -21,3 +21,4 @@ myapi.key
 xv6.out*
 .vagrant/
 submissions/
+.vscode/
\ No newline at end of file
diff --git a/kernel/defs.h b/kernel/defs.h
index c3934d4..c4d57e8 100644
--- a/kernel/defs.h
+++ b/kernel/defs.h
@@ -62,6 +62,7 @@ void            ramdiskrw(struct buf*);
 void*           kalloc(void);
 void            kfree(void *);
 void            kinit();
+void            increment_ref(uint64);
 
 // log.c
 void            initlog(int, struct superblock*);
@@ -168,6 +169,7 @@ void            uvmfree(pagetable_t, uint64);
 void            uvmunmap(pagetable_t, uint64, uint64, int);
 void            uvmclear(pagetable_t, uint64);
 uint64          walkaddr(pagetable_t, uint64);
+pte_t           *walk(pagetable_t, uint64, int);
 int             copyout(pagetable_t, uint64, char *, uint64);
 int             copyin(pagetable_t, char *, uint64, uint64);
 int             copyinstr(pagetable_t, char *, uint64, uint64);
diff --git a/kernel/file.c b/kernel/file.c
index 671e44c..706a5ce 100644
--- a/kernel/file.c
+++ b/kernel/file.c
@@ -4,7 +4,7 @@
 
 #include "types.h"
 #include "riscv.h"
-#include "defs.h"
+#include "defs.h" // contains list and buddy allocator functions
 #include "param.h"
 #include "fs.h"
 #include "spinlock.h"
@@ -16,7 +16,7 @@
 struct devsw devsw[NDEV];
 struct {
   struct spinlock lock;
-  struct file file[NFILE];
+  struct file file[NFILE]; // file array of size NFILE
 } ftable;
 
 void
@@ -31,10 +31,10 @@ filealloc(void)
 {
   struct file *f;
 
-  acquire(&ftable.lock);
-  for(f = ftable.file; f < ftable.file + NFILE; f++){
-    if(f->ref == 0){
-      f->ref = 1;
+  acquire(&ftable.lock); // Lock to ensure table isn't accessed by two diff procs at same time
+  for(f = ftable.file; f < ftable.file + NFILE; f++){ // Look for a free file slot
+    if(f->ref == 0){ // Unused file slot found
+      f->ref = 1; // Designate that we are now using this file
       release(&ftable.lock);
       return f;
     }
diff --git a/kernel/kalloc.c b/kernel/kalloc.c
index fa6a0ac..6e6f33e 100644
--- a/kernel/kalloc.c
+++ b/kernel/kalloc.c
@@ -18,6 +18,8 @@ struct run {
   struct run *next;
 };
 
+struct ref_locked ref_lock;
+
 struct {
   struct spinlock lock;
   struct run *freelist;
@@ -27,6 +29,7 @@ void
 kinit()
 {
   initlock(&kmem.lock, "kmem");
+  initlock(&ref_lock.lock, "ref_lock"); //LAB3
   freerange(end, (void*)PHYSTOP);
 }
 
@@ -35,8 +38,23 @@ freerange(void *pa_start, void *pa_end)
 {
   char *p;
   p = (char*)PGROUNDUP((uint64)pa_start);
-  for(; p + PGSIZE <= (char*)pa_end; p += PGSIZE)
+  for(; p + PGSIZE <= (char*)pa_end; p += PGSIZE) {
+    acquire(&ref_lock.lock);
+    uint64 pageindex = (uint64)p / PGSIZE;
+    ref_lock.refcount[pageindex] = 1; // init page ref count to 1 - kfree() = 0
+    release(&ref_lock.lock);
+    
     kfree(p);
+  }
+}
+
+// LAB3
+// increments ref count for the page at a specific physical address
+void increment_ref(uint64 pa) {
+  acquire(&ref_lock.lock);
+  uint64 pageindex = (uint64)pa / PGSIZE;
+  ref_lock.refcount[pageindex]++; // LAB3: set ref to 1 by incrementing by 1
+  release(&ref_lock.lock);
 }
 
 // Free the page of physical memory pointed at by v,
@@ -46,20 +64,28 @@ freerange(void *pa_start, void *pa_end)
 void
 kfree(void *pa)
 {
-  struct run *r;
+  // LAB3: check to see if no other procs are using this page
+  acquire(&ref_lock.lock);
+  uint64 pageindex = (uint64)pa / PGSIZE;
+  ref_lock.refcount[pageindex]--; // decrease ref by 1
+  release(&ref_lock.lock);
+  
+  if (ref_lock.refcount[pageindex] == 0) { // if ref == 0, no procs are using the page.
+    struct run *r;
 
-  if(((uint64)pa % PGSIZE) != 0 || (char*)pa < end || (uint64)pa >= PHYSTOP)
-    panic("kfree");
+    if(((uint64)pa % PGSIZE) != 0 || (char*)pa < end || (uint64)pa >= PHYSTOP)
+      panic("kfree");
 
-  // Fill with junk to catch dangling refs.
-  memset(pa, 1, PGSIZE);
+    // Fill with junk to catch dangling refs.
+    memset(pa, 1, PGSIZE);
 
-  r = (struct run*)pa;
+    r = (struct run*)pa;
 
-  acquire(&kmem.lock);
-  r->next = kmem.freelist;
-  kmem.freelist = r;
-  release(&kmem.lock);
+    acquire(&kmem.lock);
+    r->next = kmem.freelist;
+    kmem.freelist = r;
+    release(&kmem.lock);
+  }
 }
 
 // Allocate one 4096-byte page of physical memory.
@@ -72,11 +98,18 @@ kalloc(void)
 
   acquire(&kmem.lock);
   r = kmem.freelist;
-  if(r)
-    kmem.freelist = r->next;
+  if(r) // if 0x0, aka not null
+    kmem.freelist = r->next; // then go to the next one
   release(&kmem.lock);
 
-  if(r)
+  if(r) { // if 0x0, aka not null
     memset((char*)r, 5, PGSIZE); // fill with junk
+
+    acquire(&ref_lock.lock);
+    uint64 pageindex = (uint64)r / PGSIZE;
+    ref_lock.refcount[pageindex] = 1; // set ref to 1
+    release(&ref_lock.lock);
+  }
+
   return (void*)r;
-}
+}
\ No newline at end of file
diff --git a/kernel/param.h b/kernel/param.h
index 8e4ca98..38a63a1 100644
--- a/kernel/param.h
+++ b/kernel/param.h
@@ -12,3 +12,5 @@
 #define FSSIZE       2000  // size of file system in blocks
 #define MAXPATH      128   // maximum file path name
 #define NDISK        2
+#define PGSIZE 4096 // bytes per page
+#define PGSHIFT 12  // bits of offset within a page
diff --git a/kernel/riscv.h b/kernel/riscv.h
index f46ba59..f5078ec 100644
--- a/kernel/riscv.h
+++ b/kernel/riscv.h
@@ -332,6 +332,7 @@ sfence_vma()
 #define PTE_W (1L << 2)
 #define PTE_X (1L << 3)
 #define PTE_U (1L << 4) // 1 -> user can access
+#define PTE_COW (1L << 5) // is a Copy-On-Write mapped PTE
 
 // shift a physical address to the right place for a PTE.
 #define PA2PTE(pa) ((((uint64)pa) >> 12) << 10)
diff --git a/kernel/spinlock.h b/kernel/spinlock.h
index 90990a2..2f5d263 100644
--- a/kernel/spinlock.h
+++ b/kernel/spinlock.h
@@ -1,3 +1,6 @@
+#include "memlayout.h"
+#include "param.h"
+
 // Mutual exclusion lock.
 struct spinlock {
   uint locked;       // Is the lock held?
@@ -9,3 +12,7 @@ struct spinlock {
   uint nts;
 };
 
+extern struct ref_locked{ // LAB3
+  struct spinlock lock;
+  int refcount[PHYSTOP/PGSIZE]; // points to linked list of pages, at the head
+} ref_lock;
diff --git a/kernel/trap.c b/kernel/trap.c
index 51a7805..c41a29b 100644
--- a/kernel/trap.c
+++ b/kernel/trap.c
@@ -67,6 +67,47 @@ usertrap(void)
     syscall();
   } else if((which_dev = devintr()) != 0){
     // ok
+  } else if(r_scause() == 15) { // write page fault
+    //LAB 3
+    uint64 va_fault= PGROUNDDOWN(r_stval()); // get VA from current trapped process
+    
+    pte_t *pte_fault;
+  
+    if((pte_fault = walk(p->pagetable, va_fault, 0)) == 0) // check to see if the PA at virtual address va_fault exists
+      panic("usertrap(): pte_fault should exist\n");
+
+    if((*pte_fault & PTE_V) && (*pte_fault & PTE_U) && *pte_fault & PTE_COW) { // is a COW page
+      // for the physpage_new, turn on PTE_W
+      uint flags = PTE_FLAGS(*pte_fault);
+      flags |= PTE_W;
+      flags &= (~PTE_COW);
+
+      char *physpage_new;
+      char *pa0 = (char *)PTE2PA(*pte_fault);
+
+      if((physpage_new = kalloc()) == 0) {
+        p->killed = 1; // no available memory
+        exit(-1);
+      }
+
+      memmove(physpage_new, pa0, PGSIZE); // copy block of memory from parent's page PA to the new page of phys memory [physpage_new] we allocated
+
+      // unmap the faulting virtual page from the COW page
+      uvmunmap(p->pagetable, va_fault, PGSIZE, 0);
+
+      // decrement page reference counter of PA
+      uint64 pageindex = (uint64)pa0/PGSIZE;
+      acquire(&ref_lock.lock);
+      ref_lock.refcount[pageindex]--;
+      release(&ref_lock.lock);
+
+      // map the faulting virtual page to new page
+      if(mappages(p->pagetable, va_fault, PGSIZE, (uint64)physpage_new, flags) != 0) { // failed to map a PTE VA for our new page of phys memory
+        printf("ERR: usertrap(): failed to map virtual page\n");
+        p->killed = 1;
+        exit(-1);
+      }
+    }
   } else {
     printf("usertrap(): unexpected scause %p pid=%d\n", r_scause(), p->pid);
     printf("            sepc=%p stval=%p\n", r_sepc(), r_stval());
diff --git a/kernel/vm.c b/kernel/vm.c
index dd65184..73499d6 100644
--- a/kernel/vm.c
+++ b/kernel/vm.c
@@ -5,6 +5,7 @@
 #include "riscv.h"
 #include "defs.h"
 #include "fs.h"
+#include "spinlock.h"
 
 /*
  * the kernel's page table.
@@ -75,7 +76,7 @@ kvminithart()
 //   21..39 -- 9 bits of level-1 index.
 //   12..20 -- 9 bits of level-0 index.
 //    0..12 -- 12 bits of byte offset within the page.
-static pte_t *
+pte_t *
 walk(pagetable_t pagetable, uint64 va, int alloc)
 {
   if(va >= MAXVA)
@@ -322,20 +323,29 @@ uvmcopy(pagetable_t old, pagetable_t new, uint64 sz)
   pte_t *pte;
   uint64 pa, i;
   uint flags;
-  char *mem;
 
-  for(i = 0; i < sz; i += PGSIZE){
-    if((pte = walk(old, i, 0)) == 0)
+  for(i = 0; i < sz; i += PGSIZE){ // loops through virtual address space up to sz (the proc's memory size)
+    if((pte = walk(old, i, 0)) == 0) // check to see if the PTE at virtual address i exists
       panic("uvmcopy: pte should exist");
-    if((*pte & PTE_V) == 0)
+    if((*pte & PTE_V) == 0) // checks to see if PTE flag bit is valid/present
       panic("uvmcopy: page not present");
-    pa = PTE2PA(*pte);
-    flags = PTE_FLAGS(*pte);
-    if((mem = kalloc()) == 0)
-      goto err;
-    memmove(mem, (char*)pa, PGSIZE);
-    if(mappages(new, i, PGSIZE, (uint64)mem, flags) != 0){
-      kfree(mem);
+
+
+    pa = PTE2PA(*pte); // converts child PTE to PA
+    flags = PTE_FLAGS(*pte); // retrieves the flags from child PTE
+    flags |= PTE_COW;
+    flags &= (~PTE_W); // removes write access for child and parent PTEs
+
+    if(mappages(new, i, PGSIZE, (uint64)pa, flags) != 0) { // map parent's PA to child's VA
+      printf("ERR: uvmcopy() failed to map page");
+      goto err; // failed to map page
+    }
+
+    // increment page reference counter of parent PA
+    increment_ref(pa);
+
+    uvmunmap(old, i, PGSIZE, 0); // remove parent mapping
+    if(mappages(old, i, PGSIZE, pa, flags) != 0) { // update mapping with new flags
       goto err;
     }
   }
@@ -365,21 +375,58 @@ uvmclear(pagetable_t pagetable, uint64 va)
 int
 copyout(pagetable_t pagetable, uint64 dstva, char *src, uint64 len)
 {
-  uint64 n, va0, pa0;
+  uint64 n, va0;
 
   while(len > 0){
     va0 = PGROUNDDOWN(dstva);
-    pa0 = walkaddr(pagetable, va0);
-    if(pa0 == 0)
+    pte_t *pte_fault = walk(pagetable, va0, 0);
+    if(pte_fault == 0)
       return -1;
-    n = PGSIZE - (dstva - va0);
-    if(n > len)
-      n = len;
-    memmove((void *)(pa0 + (dstva - va0)), src, n);
 
-    len -= n;
-    src += n;
-    dstva = va0 + PGSIZE;
+    if ((*pte_fault & PTE_W) == 0) { // not writable
+      if((*pte_fault & PTE_V) && (*pte_fault & PTE_U) && *pte_fault & PTE_COW) { // is a COW page
+        // set flags
+        uint flags = PTE_FLAGS(*pte_fault);
+        flags |= PTE_W;
+        flags &= (~PTE_COW);
+
+        char *physpage_new;
+        char *pa0 = (char *)PTE2PA(*pte_fault);
+      
+        if((physpage_new = kalloc()) == 0) { // allocates a new page of phys memory
+          printf("ERR: copyout(): failed to alloc new phys memory page\n");
+          exit(-1);
+        }
+      
+        memmove(physpage_new, pa0, PGSIZE); // copy block of memory from parent's page PA to the new page of phys memory [physpage_new] we allocated
+      
+        // unmap the faulting virtual page from the COW page
+        uvmunmap(pagetable, va0, PGSIZE, 0);
+
+        // decrement page reference counter of PA
+        uint64 pageindex = (uint64)pa0/PGSIZE;
+        acquire(&ref_lock.lock);
+        ref_lock.refcount[pageindex]--;
+        release(&ref_lock.lock);
+      
+        // map the faulting virtual page to new page
+        if(mappages(pagetable, va0, PGSIZE, (uint64)physpage_new, flags) != 0) { // failed to map a PTE VA for our new page of phys memory
+          printf("ERR: usertrap(): failed to map virtual page\n");
+          exit(-1);
+        }
+      }
+    } else { //writable
+      uint64 pa0;
+      pa0 = PTE2PA(*pte_fault);
+      n = PGSIZE - (dstva - va0);
+      if(n > len)
+        n = len;
+      memmove((void *)(pa0 + (dstva - va0)), src, n);
+  
+      len -= n;
+      src += n;
+      dstva = va0 + PGSIZE;
+    }
   }
   return 0;
 }
diff --git a/user/find.c b/user/find.c
new file mode 100644
index 0000000..a673662
--- /dev/null
+++ b/user/find.c
@@ -0,0 +1,80 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+#include "kernel/fs.h"
+
+void find(char *path, char *target) {
+    int file_descriptor; // an identifier # for a file/dir, so we can lookup it's directory_entry
+    struct stat stats; // file size, permissions, timestamps, file type, and inode number.
+    // inode number: unique id for the file, used by system to locate a file's metadata
+
+    // stores path as an int into file_descriptor 
+    if((file_descriptor = open(path, 0)) < 0) { // open 0 : read_only mode
+        fprintf(2, "ERROR: find: cannot open %s\n", path);
+        return;
+    }
+
+    // get stats from input (file_descriptor)
+    if(fstat(file_descriptor, &stats) < 0) { // fstat inputs a file_descriptor to return stats, stat inputs a path
+        fprintf(2, "ERROR: find: cannot stat %s\n", path);
+        close(file_descriptor);
+        return;
+    }
+
+    // Ensure it's a directory before proceeding
+    if (stats.type != T_DIR) {
+        fprintf(2, "ERROR: find: %s is not a directory\n", path);
+        close(file_descriptor);
+        return;
+    }
+
+    char buffer[512]; // buffer 
+    char *p; // allows us to manipulate the buffer
+    struct dirent directory_entry; // a directory entry from a table containing [inum (inode number), file name]
+
+    // add onto the buffer
+    strcpy(buffer, path); // copy path to buffer
+    p = buffer + strlen(buffer); // sets pointer p to the end of the string in buffer, p is buffer now
+    *p++ = '/'; // add slash at end of p and increment pointer
+    //printf("DEBUG: Found a directory - %s\n", buffer);
+
+    // iterate through all files in directory
+    // - each read() steps through the directory entries, returning the next one
+    while(read(file_descriptor, &directory_entry, sizeof(directory_entry)) == sizeof(directory_entry)) {
+        if(directory_entry.inum == 0 
+            || strcmp(directory_entry.name, ".") == 0 
+            || strcmp(directory_entry.name, "..") == 0
+            ) // skip empty files, ".", and ".."
+            continue;
+        
+        memmove(p, directory_entry.name, DIRSIZ); // adds the current directory_entry.name to the end of the buffer
+        p[DIRSIZ] = 0; // null terminate the buffer so it prints properly
+        //printf("DEBUG: buf = %s\n", buffer);
+
+        // Get the status of the current entry
+        if (stat(buffer, &stats) < 0) {
+            fprintf(2, "ERROR: find: cannot stat %s\n", buffer);
+            continue;
+        }
+
+        //printf("DEBUG: %s matches %s?\n", directory_entry.name, target);
+        if(strcmp(directory_entry.name, target) == 0) { // check if file name matches target
+            //printf("DEBUG: MATCHED\n");
+            printf("%s\n", buffer);
+        }
+
+        if(stats.type == T_DIR) {
+            find(buffer, target); // enter next directory (recursion)
+        }
+    }
+}
+
+int main(int argc, char *argv[]) {
+  if (argc != 3) {
+    printf("Usage: find <path> <filename>\n");
+    exit();
+  }
+
+  find(argv[1], argv[2]);
+  exit();
+}
\ No newline at end of file
diff --git a/user/sleep.c b/user/sleep.c
new file mode 100644
index 0000000..e18672e
--- /dev/null
+++ b/user/sleep.c
@@ -0,0 +1,19 @@
+#include "kernel/types.h"
+#include "kernel/stat.h"
+#include "user/user.h"
+
+int main(int argc, char *argv[]) {
+    // use ticks at time measurement
+    // get start time at function call,
+        // wait(ticks)
+    // check to see if time equals start time + inputted wait time (end time)
+        // if equal or higher, continue execution
+
+    if(argc <= 1) {
+        printf("sleep: no arguments passed!\n");
+        exit();
+    }
+    
+    sleep(atoi(argv[1]));
+    exit();
+}
\ No newline at end of file
diff --git a/user/usertests.c b/user/usertests.c
index db9f680..9d71fe0 100644
--- a/user/usertests.c
+++ b/user/usertests.c
@@ -2125,7 +2125,7 @@ main(int argc, char *argv[])
     char *s;
   } tests[] = {
     {reparent2, "reparent2"},
-    {pgbug, "pgbug" },
+    //{pgbug, "pgbug" },
     {sbrkbugs, "sbrkbugs" },
     // {badwrite, "badwrite" },
     {badarg, "badarg" },
@@ -2133,7 +2133,7 @@ main(int argc, char *argv[])
     {twochildren, "twochildren"},
     {forkfork, "forkfork"},
     {forkforkfork, "forkforkfork"},
-    {argptest, "argptest"},
+    //{argptest, "argptest"},
     {createdelete, "createdelete"},
     {linkunlink, "linkunlink"},
     {linktest, "linktest"},
diff --git a/user/xargs.c b/user/xargs.c
new file mode 100644
index 0000000..16234eb
--- /dev/null
+++ b/user/xargs.c
@@ -0,0 +1,88 @@
+#include "kernel/types.h"
+#include "user/user.h"
+#include "kernel/fs.h"
+#include "kernel/param.h" // For MAXARG
+
+//TODO: for some reason the debug print statement doesnt correctly read out the initial_cmd after modifying it. 
+// might be a sign of unexpected behavior?
+
+void run_command(char **initial_cmd, char *piped_input, int initial_cmd_count) {
+    /*
+    printf("DEBUG: Running command with parameters:\n");
+    for(int i = 0; initial_cmd[i] != 0; i++) {
+        printf("DEBUG: initial_cmd[%d] - %s\n", i, initial_cmd[i]);
+    }
+    printf("DEBUG: arg - %s\n", piped_input);
+    printf("DEBUG: initial_cmd_count - %d\n", initial_cmd_count);
+    */
+
+    // adds the additional argument from the input line
+    initial_cmd[initial_cmd_count++] = piped_input;
+    initial_cmd[initial_cmd_count] = 0; // Null-terminate the argv array
+
+    /*
+    for(int i = 0; initial_cmd[i] != 0; i++) {
+            printf("DEBUG: initial_cmd[%d] - %s\n", i, initial_cmd[i]);
+    }
+    */
+
+    // fork a child process to execute the command
+    int pid = fork();
+    if (pid < 0) {
+        printf("ERROR: xargs: fork()\n");
+        exit();
+    } else if (pid == 0) {
+        // In child process, execute the command
+        //printf("DEBUG: CHILD PROCESS pid - %d\n", pid);
+        
+        exec(initial_cmd[0], initial_cmd); //(path of executable aka executable name, arguments)
+        printf("ERROR: xargs: exec()\n");
+        exit();
+    } else {
+        //printf("DEBUG: PARENT PROCESS pid - %d\n", pid);
+        // In parent process, wait for the child to complete
+        wait();
+        //printf("DEBUG: child_pid - %d\n", child_pid);
+    }
+}
+
+int main(int argc, char *argv[]) {
+    /*
+    for(int i = 0; i < argc; i++) {
+            printf("DEBUG: argv[%d] - %s\n", i, argv[i]);
+    }
+    */
+
+    if (argc < 2) { // must have atleast one argument
+        printf("Usage: xargs <command>\n");
+        exit();
+    }
+
+    // read() input lines one char at a time
+    char line[MAX_LINE];
+    int index = 0;
+    while (1) {
+        char c;
+        int bytes_read = read(0, &c, 1); // 0 for stdin
+        //printf("DEBUG: %c\n", c);
+        //printf("DEBUG: index before - %d\n", index);
+        if (bytes_read <= 0) { // End of piped input or error
+            //printf("DEBUG: End of piped input\n");
+            if (index > 0) {
+                line[index] = 0; // null terminate the line
+                run_command(argv + 1, line, argc - 1);
+            }
+            break;
+        } else if (c == '\n') {
+            //printf("DEBUG: End of line found\n");
+            line[index] = 0; // null terminate the line
+            run_command(argv + 1, line, argc - 1);
+            index = 0; // reset for the next line
+        } else if (index < MAX_LINE - 1) { // if index not at the end of the line
+            line[index++] = c; // increment index
+        }
+        //printf("DEBUG: index after - %d\n", index);
+    }
+
+    return 0;
+}
